stan/math/prim/arr/fun/array_builder.hpp

```diff
@@ -0,0 +1,47 @@
+#ifndef STAN_MATH_PRIM_ARR_FUN_ARRAY_BUILDER_HPP
+#define STAN_MATH_PRIM_ARR_FUN_ARRAY_BUILDER_HPP
+
+#include <stan/math/prim/scal/fun/promote_elements.hpp>
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Structure for building up arrays in an expression (rather than
+     * in statements) using an argumentchaining add() method and
+     * a getter method array() to return the result.
+     * Array elements are held in std::vector of type T.
+     *
+     * @tparam T type of array elements
+     */
+    template <typename T>
+    struct array_builder {
+      std::vector<T> x_;
+      array_builder() : x_() { }
+
+      /**
+       * Add one element of type S to array, promoting to type T.
+       *
+       * @param u element to add
+       * @returns this array_builder object
+       */
+      template <typename S>
+      array_builder& add(const S& u) {
+        x_.push_back(promote_elements<T, S>::promote(u));
+        return *this;
+      }
+
+      /**
+       * Getter method to return array itself.
+       *
+       * @returns std:vector<T> of composed array elements.
+       */
+      std::vector<T> array() {
```

Comment:
the array_builder object holds a reference to the array object as a private member variable.

in the generated stan c++ code, an array_builder object is instantiated and the array() method is called as the 2nd argument to the stan::math::assign function:

stan program stmt:
```
 int td_arr_int_d1_1[3] = {1, 2, 3};
```

compiled to:
```
        validate_non_negative_index("td_arr_int_d1_1", "3", 3);
        td_arr_int_d1_1 = std::vector<int>(3,int(0));
        stan::math::fill(td_arr_int_d1_1, std::numeric_limits<int>::min());
        stan::math::assign(td_arr_int_d1_1,static_cast<std::vector<int> >(stan::math::array_builder<int >().add(1).add(2).add(3).array()));
```
does this constitute copy on use?

declaring the array() method:
```
const std::vector<T>& array();
```
and then compiling and running the test models in `stan/src/test/test-models/good/array-expr/validate_array_expr_primitives` and `validate_array_expr_containers`, which have data.R files checked in for running the test models, produced no compiler or runtime errors.